spec.html

```diff
@@ -119,6 +119,12 @@ <h1>
 		<emu-note type=editor>
 			In the case of an abrupt ~throw~ completion, the type of error to be created should match the type of the abrupt throw completion record. This could be revisited when merging into the main specification. Additionally, in the case of a ~break~ or ~continue~ completion, since those are not supported, a TypeError is expected.
 		</emu-note>
+		<p>
```

Comment:
@mhofman @erights let me know if this is enough information to prevent the violation of the callable boundary to occur when a ShadowRealm is involved.

Reply:
My concern here is two-fold:
- there is no notion in the spec of what an "object reference" is, so such normative text may not hold much meaning.
- Most objects not not have a concept of belonging to a realm in the spec (in particular they don't have a `[[Realm]]` slot unlike most functions / callable objects).

That's the reason why I went for a more formal kind of wording derived from the definition of liveness in my suggestion in https://github.com/tc39/proposal-shadowrealm/issues/324#issuecomment-1145595831:

> If an execution in a shadow realm R1 oblivious of host APIs can observe the identity of an object O1, a host API Must Not allow an execution in any other realm than R1 to also observe the identity of O1. Similarly if an execution in a realm R2 can observe the identity of an object O2, a host API Must Not allow execution in any other realm than R2 that is a Shadow realm to also observe the identity of O2.

Regarding host vs implementation:

@erights wrote:
> Does "host environment" cover implementation differences? The v8 APIs of concern here are v8 implementation specific but host independent.

@caridy wrote:
> And it seems to me that "host environment" (_which typically includes objects or functions which allow obtaining input and providing output as host-defined properties of the global object_) is the right term for this definition. Maybe one of the editors can help here.

That seems right. In my mind the Host is the part that defines hooks and extensions for which the implementation is not specified in ECMA262, where implementation is for the parts covered by this spec. Since this spec text does not provide for an implementation to expose objects across the callable boundary, only host-defined behavior would if not constrained.

Reply:
@mhofman wrote:

> If an execution in a shadow realm R1 oblivious of host APIs can observe the identity of an object O1, a host API Must Not allow an execution in any other realm than R1 to also observe the identity of O1. Similarly if an execution in a realm R2 can observe the identity of an object O2, a host API Must Not allow execution in any other realm than R2 that is a Shadow realm to also observe the identity of O2.

Well, I think you're right! I was concerned about allowing a host to share objects even if their identify cannot be observed, but that's what a record/tuple is. I think if the host can keep that guarantee, they can share objects across realms. I will update the spec to go with identity rather than reference.

Reply:
Yeah if we ever have objects without identity, allowing them to cross the boundary would likely be ok. Right now records and tuples are still not objects so we're fine.

If you go with my text, or with something similar, we likely need an editor note to put in more digestible terms the intent of the normative text.

Reply:
@caridy, as I explained above, I still think that "an object from" does not mean anything in normative text. Objects do not have attached realms in the spec.

Reply:
Yes, I have expanded the spec text to reflect that, and I have expanded the editor note to provide more details based on your proposed text.

Reply:
I would have switched the 2 around, aka my text as the normative part, and your text as the editor note. My text is definitely too dense to be a clarifying note 😉

Reply:
I have never seen text with examples in the spec text... that's why I think we should keep your note as editorial, so the editors can adjust the spec text accordingly as they see fit.

Reply:
I'm not sure what you mean by examples?

The definition of liveness is structured exactly the way I suggest.

Reply:
`liveness` are all notes from what I remember. Anyhow, let's way for the editors to chime in.

Reply:
Are these questions not covered by [Hosts and Implementations](https://tc39.es/ecma262/multipage/overview.html#sec-hosts-and-implementations)?

<blockquote>

<p>A <dfn id="host" variants="hosts"><em>host</em></dfn> is an external source that further defines facilities listed in Annex <emu-xref href="#sec-host-layering-points" id="_ref_5"><a href="https://tc39.es/ecma262/multipage/host-layering-points.html#sec-host-layering-points">D</a></emu-xref> but does not further define other <emu-xref href="#implementation-defined" id="_ref_990"><a href="https://tc39.es/ecma262/multipage/overview.html#implementation-defined">implementation-defined</a></emu-xref> or <emu-xref href="#implementation-approximated" id="_ref_991"><a href="https://tc39.es/ecma262/multipage/overview.html#implementation-approximated">implementation-approximated</a></emu-xref> facilities. In informal use, a <emu-xref href="#host" id="_ref_992"><a href="https://tc39.es/ecma262/multipage/overview.html#host">host</a></emu-xref> refers to the set of all implementations, such as the set of all web browsers, that interface with this specification in the same way via Annex <emu-xref href="#sec-host-layering-points" id="_ref_6"><a href="https://tc39.es/ecma262/multipage/host-layering-points.html#sec-host-layering-points">D</a></emu-xref>. A <emu-xref href="#host" id="_ref_993"><a href="https://tc39.es/ecma262/multipage/overview.html#host">host</a></emu-xref> is often an external specification, such as WHATWG HTML (<a href="https://html.spec.whatwg.org/">https://html.spec.whatwg.org/</a>). In other words, facilities that are <emu-xref href="#host-defined" id="_ref_994"><a href="https://tc39.es/ecma262/multipage/overview.html#host-defined">host-defined</a></emu-xref> are often further defined in external specifications.</p>
    <p>A <dfn id="host-hook" variants="host hooks"><em>host hook</em></dfn> is an abstract operation that is defined in whole or in part by an external source. All <emu-xref href="#host-hook" id="_ref_995"><a href="https://tc39.es/ecma262/multipage/overview.html#host-hook">host hooks</a></emu-xref> must be listed in Annex <emu-xref href="#sec-host-layering-points" id="_ref_7"><a href="https://tc39.es/ecma262/multipage/host-layering-points.html#sec-host-layering-points">D</a></emu-xref>. A <emu-xref href="#host-hook" id="_ref_996"><a href="https://tc39.es/ecma262/multipage/overview.html#host-hook">host hook</a></emu-xref> must conform to at least the following requirements:</p>
    <ul>
      <li>It must return either a <emu-xref href="#sec-completion-record-specification-type" id="_ref_997"><a href="https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-completion-record-specification-type">normal completion</a></emu-xref> or a <emu-xref href="#sec-completion-record-specification-type" id="_ref_998"><a href="https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-completion-record-specification-type">throw completion</a></emu-xref>.</li>
    </ul>
    <p>A <dfn id="host-defined"><em>host-defined</em></dfn> facility is one that defers its definition to an external source without further qualification and is listed in Annex <emu-xref href="#sec-host-layering-points" id="_ref_8"><a href="https://tc39.es/ecma262/multipage/host-layering-points.html#sec-host-layering-points">D</a></emu-xref>. Implementations that are not <emu-xref href="#host" id="_ref_999"><a href="https://tc39.es/ecma262/multipage/overview.html#host">hosts</a></emu-xref> may also provide definitions for <emu-xref href="#host-defined" id="_ref_1000"><a href="https://tc39.es/ecma262/multipage/overview.html#host-defined">host-defined</a></emu-xref> facilities.</p>
    <p>A <dfn id="host-environment" variants="host environments"><em>host environment</em></dfn> is a particular choice of definition for all <emu-xref href="#host-defined" id="_ref_1001"><a href="https://tc39.es/ecma262/multipage/overview.html#host-defined">host-defined</a></emu-xref> facilities. A <emu-xref href="#host-environment" id="_ref_1002"><a href="https://tc39.es/ecma262/multipage/overview.html#host-environment">host environment</a></emu-xref> typically includes objects or functions which allow obtaining input and providing output as <emu-xref href="#host-defined" id="_ref_1003"><a href="https://tc39.es/ecma262/multipage/overview.html#host-defined">host-defined</a></emu-xref> properties of the <emu-xref href="#sec-global-object" id="_ref_1004"><a href="https://tc39.es/ecma262/multipage/global-object.html#sec-global-object">global object</a></emu-xref>.</p>

</blockquote>

Reply:
Hmm. In that case, "host" does not cover "implementation" and the prohibition must cover both.